misk/src/main/kotlin/misk/clustering/DefaultCluster.kt

```diff
@@ -0,0 +1,156 @@
+  }
+
+  /** Adds a new cluster watch */
+  override fun watch(watch: ClusterWatch) {
+    actions.add { watches ->
+      watches.add(watch)
+      watch(Cluster.Changes(snapshot))
+    }
+  }
+
+  /** Triggers a change to the cluster in response members becoming ready or not ready */
+  fun clusterChanged(
+    membersBecomingReady: Set<Cluster.Member> = setOf(),
+    membersBecomingNotReady: Set<Cluster.Member> = setOf()
+  ) {
+    actions.add { watches ->
+      val changeTracker = ClusterChangeTracker(this)
+      membersBecomingReady.forEach { changeTracker.memberReady(it) }
+      membersBecomingNotReady.forEach { changeTracker.memberNotReady(it) }
+      val changes = changeTracker.computeChanges()
+
+      if (changes.hasDiffs) {
+        snapshotRef.set(changes.snapshot)
+        watches.forEach {
+          try {
+            it(changes)
+          } catch (th: Throwable) {
+            log.error(th) { "error triggering watch in response to cluster change" }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Triggers a callback once all of the actions on the queue have been processed. Useful
+   * for writing deterministic tests
+   */
+  internal fun syncPoint(callback: () -> Unit) {
+    actions.add { _ -> callback() }
+  }
+
+  override fun newResourceMapper(readyMembers: Set<Cluster.Member>) =
+      newResourceMapperFn(readyMembers)
+
+  /**
+   * [ClusterChangeTracker] is an internal helper class used to track diffs to a cluster as
+   * members become ready or not ready
+   */
+  private class ClusterChangeTracker(
+    private val cluster: Cluster,
+    private val snapshot: Cluster.Snapshot = cluster.snapshot
+  ) {
+    private val readyMembers = snapshot.readyMembers.map { it.name to it }.toMap().toMutableMap()
+    private val membersAdded = mutableSetOf<Cluster.Member>()
+    private val membersRemoved = mutableSetOf<Cluster.Member>()
+
+    fun memberReady(member: Cluster.Member) {
+      if (readyMembers[member.name] == member) {
+        // We already know about this as a ready member and all of its properties are the same,
+        // so no need to apply any changes
+        return
+      }
+
+      readyMembers[member.name] = member
+      membersAdded += member
+    }
+
+    fun memberNotReady(member: Cluster.Member) {
+      if (readyMembers.remove(member.name) != null) {
+        membersRemoved += member
+      }
+    }
+
+    fun computeChanges(): Cluster.Changes {
+      val newReadyMembers = readyMembers.values.toSet()
+      return Cluster.Changes(Cluster.Snapshot(
+          self = snapshot.self,
+          selfReady = readyMembers.containsKey(snapshot.self.name),
+          readyMembers = newReadyMembers,
+          resourceMapper = cluster.newResourceMapper(newReadyMembers)
+      ), added = membersAdded.toSet(), removed = membersRemoved.toSet())
+    }
+  }
+
+  companion object {
+    val log = getLogger<DefaultCluster>()
```

Comment:
best place to declare these? I’m indifferent but I know we also do `logger` up top. I think I prefer this because of how it looks in the bytecodes

Reply:
Yeah, I'm just not really a fan of free-standing private variables, and particularly don't like having the first thing you see in the file be a minor implementation var. 

I guess looking at the bytecodes is an Android thing? I really don't care what server bytecodes look like, unless I'm writing some obj-asm code.

Reply:
Yeah, I’m thinking too hard about Android stuff where minimizing the number of `.class` files in the .jar is a thing.